hphp/runtime/ext/hsl/ext_hsl_os.cpp

```diff
@@ -408,6 +419,23 @@ struct HSLFileDescriptor {
     }
   }
 
+  // Called from HSL_os_poll_async with the WaitHandle it creates.
+  void addAsyncPoll(Object obj) {
+    // Remove any already finished ones. We could do this elsewhere but there
```

Comment:
probably good to talk to @jano or @alexeyt  about this; I think we need these to end up being cleaned up even if there isn't another one added.

This could end up meaning that FileAwaits need to be aware of FileDescriptors, or perhaps poll_async makes a distinct wrapper awaitable that's basically a try/finally around the FileAwait

This also ends up keeping a bunch of stack data (especially if an exception is thrown) until the next one

Reply:
My other idea was to pass an "on_destruct" callback to FileAwait. By "wrapper" do you mean a new type of WaitHandle?

Reply:
Yep, I was thinking a new kind of WaitHandle.